File: ios/MochiApp/Components/CreateGlyphs.swift
import SwiftUI

/// Marks on the Create Custom Theme frame (docs/figma/4.png) that no SF Symbol comes close enough
/// to. Everything else on that screen uses a system symbol; these four are drawn because the
/// design's version is either a different construction (the twelve-wedge colour wheel, the
/// hexagon) or has no system equivalent at all (the floppy disc, the keycap).

/// The twelve-wedge colour wheel inside the BACKGROUND card's "Colors" chip. Figma draws a flat
/// pie — twelve equal 30° wedges running red → orange → yellow → green → cyan → blue → violet →
/// magenta and back — not a continuous conic sweep, so the wedge boundaries stay hard.
struct ColorWheelGlyph: View {
    /// Read off the wheel in docs/figma/4.png wedge by wedge, starting at the one just clockwise
    /// of twelve o'clock and going clockwise.
    private static let wedges: [Color] = [
        Color(red: 0.98, green: 0.85, blue: 0.09),  // yellow
        Color(red: 0.97, green: 0.66, blue: 0.11),  // amber
        Color(red: 0.95, green: 0.45, blue: 0.13),  // orange
        Color(red: 0.90, green: 0.22, blue: 0.18),  // red
        Color(red: 0.83, green: 0.17, blue: 0.30),  // crimson
        Color(red: 0.66, green: 0.20, blue: 0.47),  // magenta
        Color(red: 0.47, green: 0.24, blue: 0.60),  // violet
        Color(red: 0.22, green: 0.31, blue: 0.65),  // indigo
        Color(red: 0.13, green: 0.47, blue: 0.55),  // teal
        Color(red: 0.14, green: 0.58, blue: 0.38),  // green
        Color(red: 0.30, green: 0.68, blue: 0.25),  // grass
        Color(red: 0.58, green: 0.76, blue: 0.20)   // lime
    ]

    var body: some View {
        GeometryReader { geo in
            let r = min(geo.size.width, geo.size.height) / 2
            let c = CGPoint(x: geo.size.width / 2, y: geo.size.height / 2)
            ZStack {
                ForEach(Array(Self.wedges.enumerated()), id: \.offset) { index, color in
                    Path { path in
                        path.move(to: c)
                        path.addArc(
                            center: c,
                            radius: r,
                            startAngle: .degrees(Double(index) * 30 - 90),
                            endAngle: .degrees(Double(index + 1) * 30 - 90),
                            clockwise: false
                        )
                        path.closeSubpath()
                    }
                    .fill(color)
                }
            }
        }
    }
}

/// The fourth key-shape option. SwiftUI has no hexagon primitive and SF's `hexagon` is a rounded,
/// point-up hex; Figma's is flat-topped with sharp vertices and its points on the horizontal axis.
struct HexagonShape: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        let w = rect.width, h = rect.height
        // Flat top and bottom edges spanning the middle half of the width, points at mid-height.
        let points = [
            CGPoint(x: rect.minX + w * 0.25, y: rect.minY),
            CGPoint(x: rect.minX + w * 0.75, y: rect.minY),
            CGPoint(x: rect.maxX, y: rect.midY),
            CGPoint(x: rect.minX + w * 0.75, y: rect.maxY),
            CGPoint(x: rect.minX + w * 0.25, y: rect.maxY),
            CGPoint(x: rect.minX, y: rect.midY)
        ]
        _ = h
        path.addLines(points)
        path.closeSubpath()
        return path
    }
}

/// The Save Draft button's mark: a classic floppy disc, stroked not filled — outer body with the
/// write-protect corner clipped off, the shutter across the top and the label panel below it.
/// SF's nearest options (`square.and.arrow.down`, `externaldrive`) read as entirely different
/// objects at this size.
struct FloppyGlyph: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        let w = rect.width, h = rect.height
        let corner = w * 0.22          // the clipped top-left corner

        path.move(to: CGPoint(x: rect.minX + corner, y: rect.minY))
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
        path.addLine(to: CGPoint(x: rect.minX, y: rect.maxY))
        path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + corner))
        path.closeSubpath()

        // Shutter — the metal slider across the disc's top third.
        path.addRect(CGRect(x: rect.minX + w * 0.26, y: rect.minY + h * 0.10,
                            width: w * 0.48, height: h * 0.28))
        // Label panel — the paper sticker across the bottom.
        path.addRect(CGRect(x: rect.minX + w * 0.20, y: rect.minY + h * 0.52,
                            width: w * 0.60, height: h * 0.48))
        return path
    }
}

/// The "Keys" tab mark: a solid rounded keycap with a four-point sparkle punched out of it.
struct KeycapGlyph: View {
    var body: some View {
        GeometryReader { geo in
            let w = geo.size.width, h = geo.size.height
            RoundedRectangle(cornerRadius: h * 0.28, style: .continuous)
                .overlay(
                    Image(systemName: "sparkle")
                        .resizable()
                        .scaledToFit()
                        .frame(width: w * 0.42, height: w * 0.42)
                        .foregroundStyle(.white)
                )
        }
    }
}

File: android/app/src/main/java/com/mochi/keyboard/components/CreateGlyphs.kt
package com.mochi.keyboard.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.drawscope.Fill
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

/**
 * Ported from ios/MochiApp/Components/CreateGlyphs.swift — marks on the Create Custom Theme frame
 * (docs/figma/4.png) no Material icon comes close enough to.
 */

/** The twelve-wedge flat color wheel inside the BACKGROUND card's "Colors" chip — hard wedge
 * boundaries, not a continuous conic sweep. Read off figma/4.png wedge by wedge starting just
 * clockwise of twelve o'clock. */
@Composable
fun ColorWheelGlyph(modifier: Modifier = Modifier) {
    val wedges = remember12Wedges()
    Canvas(modifier = modifier) {
        val r = min(size.width, size.height) / 2
        val c = Offset(size.width / 2, size.height / 2)
        wedges.forEachIndexed { index, color ->
            val path = Path().apply {
                moveTo(c.x, c.y)
                arcTo(
                    rect = androidx.compose.ui.geometry.Rect(c.x - r, c.y - r, c.x + r, c.y + r),
                    startAngleDegrees = index * 30f - 90f,
                    sweepAngleDegrees = 30f,
                    forceMoveTo = false
                )
                close()
            }
            drawPath(path, color, style = Fill)
        }
    }
}

private fun remember12Wedges(): List<Color> = listOf(
    Color(0xFFFAD917), Color(0xFFF7A81C), Color(0xFFF2731F),
    Color(0xFFE6382E), Color(0xFFD32B4C), Color(0xFFA83377),
    Color(0xFF783D99), Color(0xFF384FA6), Color(0xFF22788C),
    Color(0xFF249461), Color(0xFF4CAD40), Color(0xFF94C233)
)

/** SwiftUI has no hexagon primitive; SF's `hexagon` is rounded/point-up — this is flat-topped with
 * sharp vertices, points on the horizontal axis. Used as a clip/background Shape. */
class HexagonShape : Shape {
    override fun createOutline(size: Size, layoutDirection: LayoutDirection, density: Density): Outline {
        val w = size.width
        val h = size.height
        val path = Path().apply {
            moveTo(w * 0.25f, 0f)
            lineTo(w * 0.75f, 0f)
            lineTo(w, h / 2f)
            lineTo(w * 0.75f, h)
            lineTo(w * 0.25f, h)
            lineTo(0f, h / 2f)
            close()
        }
        return Outline.Generic(path)
    }
}

/** The Save Draft button's mark: a stroked classic floppy disc — clipped top-left corner, shutter
 * across the top, label panel below. SF's `square.and.arrow.down` reads as an entirely different
 * object at this size. */
@Composable
fun FloppyGlyph(modifier: Modifier = Modifier, color: Color = Color.Black, strokeWidth: Float = 1.5f) {
    Canvas(modifier = modifier) {
        val w = size.width
        val h = size.height
        val corner = w * 0.22f
        val body = Path().apply {
            moveTo(corner, 0f)
            lineTo(w, 0f)
            lineTo(w, h)
            lineTo(0f, h)
            lineTo(0f, corner)
            close()
        }
        val stroke = Stroke(strokeWidth)
        drawPath(body, color, style = stroke)
        drawRect(color, topLeft = Offset(w * 0.26f, h * 0.10f), size = Size(w * 0.48f, h * 0.28f), style = stroke)
        drawRect(color, topLeft = Offset(w * 0.20f, h * 0.52f), size = Size(w * 0.60f, h * 0.48f), style = stroke)
    }
}

/** The "Keys" tab mark: a solid rounded keycap with a four-point sparkle punched out of it. */
@Composable
fun KeycapGlyph(modifier: Modifier = Modifier, fillColor: Color = Color(0xFF9C28B1), sparkleColor: Color = Color.White) {
    Canvas(modifier = modifier) {
        val w = size.width
        val h = size.height
        drawRoundRect(
            color = fillColor,
            size = size,
            cornerRadius = androidx.compose.ui.geometry.CornerRadius(h * 0.28f, h * 0.28f)
        )
        val sparkleSize = w * 0.42f
        val sparkleTopLeft = Offset((w - sparkleSize) / 2, (h - sparkleSize) / 2)
        drawSparkle(sparkleColor, sparkleTopLeft, sparkleSize)
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawSparkle(color: Color, topLeft: Offset, s: Float) {
    val c = Offset(topLeft.x + s / 2, topLeft.y + s / 2)
    val r = s / 2
    val waist = r * 0.24f
    val path = Path().apply {
        moveTo(c.x, c.y - r)
        quadraticTo(c.x + waist, c.y - waist, c.x + r, c.y)
        quadraticTo(c.x + waist, c.y + waist, c.x, c.y + r)
        quadraticTo(c.x - waist, c.y + waist, c.x - r, c.y)
        quadraticTo(c.x - waist, c.y - waist, c.x, c.y - r)
        close()
    }
    drawPath(path, color, style = Fill)
}
